util/webhook/validator.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"net/http"
+
+	goerrors "errors"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	v1 "k8s.io/api/admission/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// Validator defines functions for validating an operation.
+type Validator interface {
+	runtime.Object
+	ValidateCreate(client client.Client) error
+	ValidateUpdate(old runtime.Object, client client.Client) error
+	ValidateDelete(client client.Client) error
+}
+
+// NewValidatingWebhook creates a new Webhook for validating the provided type.
+func NewValidatingWebhook(validator Validator, client client.Client) *admission.Webhook {
+	return &admission.Webhook{
+		Handler: &validatingHandler{
+			validator: validator,
+			Client:    client,
+		},
+	}
+}
+
+type validatingHandler struct {
+	validator Validator
+	Client    client.Client
+	decoder   *admission.Decoder
+}
+
+var _ admission.DecoderInjector = &validatingHandler{}
+
+// InjectDecoder injects the decoder into a validatingHandler.
+func (h *validatingHandler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+// Handle handles admission requests.
+func (h *validatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if h.validator == nil {
+		panic("validator should never be nil")
+	}
+
+	// Get the object in the request
+	obj := h.validator.DeepCopyObject().(Validator)
+	if req.Operation == v1.Create {
+		err := h.decoder.Decode(req, obj)
```

Comment:
is there a reason to use decode(req, obj) here but decodeRaw(req.Object, obj) below? I see controller-runtime does the same, but decode(req, obj) only does a length check before calling decodeRaw() anyway: https://github.com/kubernetes-sigs/controller-runtime/blob/1e4d87c9f9e15e4a58bb81909dd787f30ede7693/pkg/webhook/admission/decode.go#L42-L48

fine as-is, just was wondering.

Reply:
@LochanRn do you have an answer for the above?

Reply:
@CecileRobertMichon 
I picked this up from the container-runtime and this looks reasonable enough I guess.
https://github.com/kubernetes-sigs/controller-runtime/blob/1e4d87c9f9e15e4a58bb81909dd787f30ede7693/pkg/webhook/admission/decode.go#L39-L42

Reply:
yeah just to be clear this is totally fine, I suspect someone in CR didn't think too carefully about those few lines